docs/docs/reference/features-classification.md

```diff
@@ -0,0 +1,202 @@
+- [Erased Terms](https://dotty.epfl.ch/docs/reference/other-new-features/erased-terms.html) provide a general mechanism for compile-time-only computations.
+
+**Status: not yet settled**
+
+We know we need a practical replacement for current macros. The features listed above are very promising in that respect, but we need more complete implementations and more use cases to reach a final verdict.
+
+**Migration cost: very high**
+
+Existing macro libraries will have to be rewritten from the ground up. In many cases the rewritten libraries will turn out to be simpler and more robust than the old ones, but that does not relieve one of the cost of the rewrites. It's currently unclear to what degree users of macro libraries will be affected. We aim to provide sufficient functionality so that core macros can be re-implemented fully, but given the vast feature set of the various macro extensions to Scala 2 it is difficult to arrive at a workable limitation of scope.
+
+## Changes to Type Checking and Inference
+
+The Scala 3 compiler uses a new algorithm for type inference, which relies on
+a general subtype constraint solver. The new algorithm often
+[works better than the old](https://contributors.scala-lang.org/t/better-type-inference-for-scala-send-us-your-problematic-cases/2410), but there are inevitably situations where the results of both algorithms differ, leading to errors diagnosed by Scala 3 for programs that the Scala 2 compiler accepts.
+
+**Status: essential**
+
+The new type-checking and inference algorithms are the essential core of the new compiler. They cannot be reverted without dropping the whole implementation of Scala 3.
+
+**Migration cost: high**
+
+Some existing programs will break and, given the complex nature of type inference, it will not always be clear what change caused the breakage and how to fix it.
+
+In our experience, macros and changes in type and implicit argument inference together cause the large majority of problems encountered when porting existing code to Scala 3. The latter source of problems could be addressed systematically by a tool that added all inferred types and implicit arguments to a Scala 2 source code file. Most likely such a tool would be implemented as a Scala 2 compiler plugin. The resulting code would have a greatly increased likelihood to compile under Scala 3, but would often be bulky to the point of being unreadable. A second part of the rewriting tool should then selectively and iteratively remove type and implicit annotations that were synthesized by the first part as long as they compile under Scala 3. This second part could be implemented as a program that invokes the Scala 3 compiler `dotc` programmatically.
+
+Several people have proposed such a tool for some time now. I believe it is time we find the will and the resources to actually implement it.
```

Comment:
Maybe one approach to this would be:

1. As stated, a tool that rewrites all inferred types to explicitly annotated types, but for both scala 2 and scala 3
2. Use your VCS to record the result of running this both on Scala 2 and 3, and compare the results. Perhaps a merge tool with the right inputs would be a good way to streamline the process of reverting all identical added types, and highlighting the cases where they were not identical. For instance if you would have two branches where the first commit has the respective types and the next commit does not, it might find a conflict between a change A->C vs. B->C (where C is no explicit type, and A and B are the respective generated explicit type). On the other hand, wherever the generated types are the same, there would be no conflict and it would successfully apply the diff reverting to non-explicit types.

Of course, you would have to ensure that there aren't red herrings. For instance you might need to output dotty type intersections using `with`, and use the same logic for when to output types in infix format.

I'm not sure enough of the mechanics of git merge conflicts, but this might simplify things a lot, at least for implementers.

Reply:
That's an interesting technique. I had not thought of using a merge conflicts to pinpoint problems.

Btw I set out to write this document to have some data as a basis for replying to your post on Scala Contributors. I am now too tired to write a coherent reply, but will do so tomorrow.

Reply:
Yeah I figured :grinning: 

It's definitely appreciated!